### ISSUE TITLE
IndexError: list index out of range

### ISSUE THREAD

**Issue by random-tree** [metal3d/keras-video-generators#13]
Hi Patrice, thank you for sharing your code.

When i follow the exact code from you medium article

https://medium.com/smileinnovation/training-neural-network-with-image-sequence-an-example-with-video-as-input-c3407f7a0b0f

i get the following error, when i am training the model, would it be possible for you to share the notebook you created to write the medium. Thanks

IndexError: list index out of range

Thanks for your help.

**Comment by metal3d** [comment]
Hi, it's not esay to find the problem without the complete trace of the error. Could you please paste the entire stack ?

**Comment by random-tree** [comment]
Sorry about that, here is the complete error msg. Thanks 

Epoch 1/50
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-8-563749f27a65> in <module>
     13     verbose=1,
     14     epochs=EPOCHS,
---> 15     callbacks=callbacks
     16 )

~\AppData\Local\Continuum\anaconda3\lib\site-packages\keras\legacy\interfaces.py in wrapper(*args, **kwargs)
     89                 warnings.warn('Update your `' + object_name + '` call to the ' +
     90                               'Keras 2 API: ' + signature, stacklevel=2)
---> 91             return func(*args, **kwargs)
     92         wrapper._original_function = func
     93         return wrapper

~\AppData\Local\Continuum\anaconda3\lib\site-packages\keras\engine\training.py in fit_generator(self, generator, steps_per_epoch, epochs, verbose, callbacks, validation_data, validation_steps, validation_freq, class_weight, max_queue_size, workers, use_multiprocessing, shuffle, initial_epoch)
   1730             use_multiprocessing=use_multiprocessing,
   1731             shuffle=shuffle,
-> 1732             initial_epoch=initial_epoch)
   1733 
   1734     @interfaces.legacy_generator_methods_support

~\AppData\Local\Continuum\anaconda3\lib\site-packages\keras\engine\training_generator.py in fit_generator(model, generator, steps_per_epoch, epochs, verbose, callbacks, validation_data, validation_steps, validation_freq, class_weight, max_queue_size, workers, use_multiprocessing, shuffle, initial_epoch)
    183             batch_index = 0
    184             while steps_done < steps_per_epoch:
--> 185                 generator_output = next(output_generator)
    186 
    187                 if not hasattr(generator_output, '__len__'):

~\AppData\Local\Continuum\anaconda3\lib\site-packages\keras\utils\data_utils.py in get(self)
    623         except Exception:
    624             self.stop()
--> 625             six.reraise(*sys.exc_info())
    626 
    627 

~\AppData\Local\Continuum\anaconda3\lib\site-packages\six.py in reraise(tp, value, tb)
    691             if value.__traceback__ is not tb:
    692                 raise value.with_traceback(tb)
--> 693             raise value
    694         finally:
    695             value = None

~\AppData\Local\Continuum\anaconda3\lib\site-packages\keras\utils\data_utils.py in get(self)
    608                 try:
    609                     future = self.queue.get(block=True)
--> 610                     inputs = future.get(timeout=30)
    611                 except mp.TimeoutError:
    612                     idx = future.idx

~\AppData\Local\Continuum\anaconda3\lib\multiprocessing\pool.py in get(self, timeout)
    655             return self._value
    656         else:
--> 657             raise self._value
    658 
    659     def _set(self, i, obj):

~\AppData\Local\Continuum\anaconda3\lib\multiprocessing\pool.py in worker(inqueue, outqueue, initializer, initargs, maxtasks, wrap_exception)
    119         job, i, func, args, kwds = task
    120         try:
--> 121             result = (True, func(*args, **kwds))
    122         except Exception as e:
    123             if wrap_exception and func is not _helper_reraises_exception:

~\AppData\Local\Continuum\anaconda3\lib\site-packages\keras\utils\data_utils.py in get_index(uid, i)
    404         The value at index `i`.
    405     """
--> 406     return _SHARED_SEQUENCES[uid][i]
    407 
    408 

~\AppData\Local\Continuum\anaconda3\lib\site-packages\keras_video\generator.py in __getitem__(self, index)
    296 
    297             video = self.files[i]
--> 298             classname = self._get_classname(video)
    299 
    300             # create a label array and set 1 to the right column

~\AppData\Local\Continuum\anaconda3\lib\site-packages\keras_video\generator.py in _get_classname(self, video)
    325         for src, dest in RE_PATH_REPLACE.items():
    326             pattern = pattern.replace(src, dest)
--> 327         classname = re.findall(pattern, video)[0]
    328         return classname
    329 

IndexError: list index out of range

**Comment by metal3d** [comment]
OK, the problem seems to come from class detection. Are you sure that:
- the generator find some files / classes ? (you should have some information when you create the generator, it gives number of classes and files found)
- are you using Windows ? In that case, the issue #14 can explain the problem (I'm currently working on a fix)

**Comment by random-tree** [comment]
Yes, the generator was able to find classes. 
Yes i am using windows.

class dribble, validation count: 29, train count: 116
class golf, validation count: 21, train count: 84
Total data: 2 classes for 200 files for train

**Comment by metal3d** [comment]
OK, let me some times to install legacy Windows on virtual machine and test my fix.
If you really need it right now, find the file named "generator.py" in your installation (under `keras_video` directory, in site packages) and add a line in the `RE_PATH_REPLACE` variable to have `    r'\\': r'\\\\',`

The complete var should be:
```python
RE_PATH_REPLACE = {
    '.': r'\.',
    '*': r'.*',
    '?': r'\?',
    '!': r'\!',
    '+': r'\+',
    r'\\': r'\\\\',
    os.path.sep*2: os.path.sep,
    '{classname}': r'(.*?)'
}
```

It should replace backslashes correctly. If it's ok for you, I will apply the fix on the next release and fix #14 also.

**Comment by metal3d** [comment]
Woops, inverted lines, try to add the line **after** os path sep:

```python
RE_PATH_REPLACE = {
    '.': r'\.',
    '*': r'.*',
    '?': r'\?',
    '!': r'\!',
    '+': r'\+',
    os.path.sep*2: os.path.sep,
    r'\\': r'\\\\',
    '{classname}': r'(.*?)'
}
```

**Comment by random-tree** [comment]
Thanks. I tried it but it didnt work

Epoch 1/50
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-8-676487a2fdaa> in <module>
     13     verbose=1,
     14     epochs=EPOCHS,
---> 15     callbacks=callbacks
     16 )

~\AppData\Local\Continuum\anaconda3\lib\site-packages\keras\legacy\interfaces.py in wrapper(*args, **kwargs)
     89                 warnings.warn('Update your `' + object_name + '` call to the ' +
     90                               'Keras 2 API: ' + signature, stacklevel=2)
---> 91             return func(*args, **kwargs)
     92         wrapper._original_function = func
     93         return wrapper

~\AppData\Local\Continuum\anaconda3\lib\site-packages\keras\engine\training.py in fit_generator(self, generator, steps_per_epoch, epochs, verbose, callbacks, validation_data, validation_steps, validation_freq, class_weight, max_queue_size, workers, use_multiprocessing, shuffle, initial_epoch)
   1730             use_multiprocessing=use_multiprocessing,
   1731             shuffle=shuffle,
-> 1732             initial_epoch=initial_epoch)
   1733 
   1734     @interfaces.legacy_generator_methods_support

~\AppData\Local\Continuum\anaconda3\lib\site-packages\keras\engine\training_generator.py in fit_generator(model, generator, steps_per_epoch, epochs, verbose, callbacks, validation_data, validation_steps, validation_freq, class_weight, max_queue_size, workers, use_multiprocessing, shuffle, initial_epoch)
    183             batch_index = 0
    184             while steps_done < steps_per_epoch:
--> 185                 generator_output = next(output_generator)
    186 
    187                 if not hasattr(generator_output, '__len__'):

~\AppData\Local\Continuum\anaconda3\lib\site-packages\keras\utils\data_utils.py in get(self)
    623         except Exception:
    624             self.stop()
--> 625             six.reraise(*sys.exc_info())
    626 
    627 

~\AppData\Local\Continuum\anaconda3\lib\site-packages\six.py in reraise(tp, value, tb)
    691             if value.__traceback__ is not tb:
    692                 raise value.with_traceback(tb)
--> 693             raise value
    694         finally:
    695             value = None

~\AppData\Local\Continuum\anaconda3\lib\site-packages\keras\utils\data_utils.py in get(self)
    608                 try:
    609                     future = self.queue.get(block=True)
--> 610                     inputs = future.get(timeout=30)
    611                 except mp.TimeoutError:
    612                     idx = future.idx

~\AppData\Local\Continuum\anaconda3\lib\multiprocessing\pool.py in get(self, timeout)
    655             return self._value
    656         else:
--> 657             raise self._value
    658 
    659     def _set(self, i, obj):

~\AppData\Local\Continuum\anaconda3\lib\multiprocessing\pool.py in worker(inqueue, outqueue, initializer, initargs, maxtasks, wrap_exception)
    119         job, i, func, args, kwds = task
    120         try:
--> 121             result = (True, func(*args, **kwds))
    122         except Exception as e:
    123             if wrap_exception and func is not _helper_reraises_exception:

~\AppData\Local\Continuum\anaconda3\lib\site-packages\keras\utils\data_utils.py in get_index(uid, i)
    404         The value at index `i`.
    405     """
--> 406     return _SHARED_SEQUENCES[uid][i]
    407 
    408 

<ipython-input-1-eb7a0af272ac> in __getitem__(self, index)
    296 
    297             video = self.files[i]
--> 298             classname = self._get_classname(video)
    299 
    300             # create a label array and set 1 to the right column

<ipython-input-1-eb7a0af272ac> in _get_classname(self, video)
    325         for src, dest in RE_PATH_REPLACE.items():
    326             pattern = pattern.replace(src, dest)
--> 327         classname = re.findall(pattern, video)[0]
    328         return classname
    329 

IndexError: list index out of range

**Comment by metal3d** [comment]
I think I found a way in the commit 207dc93 I just pushed. But you will need to not use "." to set the current directory. If your data is in current subdirectory, use `you\path` and not `.\you\path`.

You will need my fix, I don't make a release for now as the fix is not clean (and not tested enough). 

You can try to apply the current generator.py file in your site package for now, it should work. Then I will make a proper fix and a new release.

Note: I will close #14 that is a duplicate of your issue.

One more time, thanks for your report

**Comment by metal3d** [comment]
Ad for #14, the issue is resolved in v1.0.10 - you can now upgrade the package.

Thansk for the report.
